client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/impl/EventDispatcher.java

```diff
@@ -0,0 +1,91 @@
+package org.infinispan.client.hotrod.event.impl;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import org.infinispan.client.hotrod.impl.transport.netty.ChannelRecord;
+import org.infinispan.client.hotrod.impl.transport.netty.HintedReplayingDecoder;
+import org.infinispan.client.hotrod.logging.Log;
+import org.infinispan.client.hotrod.logging.LogFactory;
+import org.infinispan.commons.marshall.WrappedByteArray;
+import org.infinispan.commons.util.Util;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+
+abstract class EventDispatcher<T> extends HintedReplayingDecoder<Void> {
+   static final Log log = LogFactory.getLog(EventDispatcher.class);
+   static final boolean trace = log.isTraceEnabled();
+   static final AtomicReferenceFieldUpdater<EventDispatcher, DispatcherStatus> statusUpdater
+         = AtomicReferenceFieldUpdater.newUpdater(EventDispatcher.class, DispatcherStatus.class, "status");
+
+   final ClientListenerNotifier notifier;
+   final Object listener;
+   final byte[] listenerId;
+   final Channel channel;
+   final String cacheName;
+   volatile DispatcherStatus status = DispatcherStatus.STOPPED;
+
+   EventDispatcher(ClientListenerNotifier notifier, String cacheName, Object listener, byte[] listenerId, Channel channel) {
+      this.notifier = notifier;
+      this.listener = listener;
+      this.listenerId = listenerId;
+      this.channel = channel;
+      this.cacheName = cacheName;
+   }
+
+   @Override
+   protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> out) throws Exception {
+      T event;
+      try {
+         event = readEvent(buf);
```

Comment:
Normally with netty you would want to read the event and then send that down the pipeline. This way you can disconnect the reading and invocation of the event into different execution handler or w/e is needed. Maybe the code isn't as conducive to this though.

Reply:
The event type that is decoded is tied to the way the event is executed. Therefore I would have two decoding classes, and two executing classes, logically coupled but theoretically interchangeable. Could you tell what are the pros? (do you expect more executing classes for the event in the future?)

Reply:
The hotrod server does this for a few different reasons.

1. Optionally installed authentication so we don't execute if that operation is blocked.
2. Optionally log details about the operation (including how long decoding and execution took)
2. Execution of some operations on a different thread. We don't want to run an operation that will block the netty thread as we could starve reads from the socket.

The last one seems like it might be important for this since we are invoking user code and that could be doing anything. If I remember correctly the events are all sent back through the same socket for all listeners for a single client.

Reply:
@wburns But when you push something further to the pipeline, you still do not specify who should handle the event, do you? It's still on the event loop thread. In this case the notifier is called in the implementation, and that can forward it to the threadpool.

Reply:
Not sure who you is in this case :) The decoder doesn't really care who handles it, it just decodes. A later Handler would have to do something with it.

The hotrod server has 2 handlers later in the pipeline that do almost the same thing (LocalContextHandler and ContextHandler) [1]. The former immediately invokes the command and sends the response (for commands that don't block) and the latter submits this on an executor to be handled in a different thread. (note these guys don't send anything else down the pipeline except exceptions).

Also you can extend `SimpleChannelInboundHandler` from netty and type it to the expected object class. For example we do that at [2] but override the matcher to be a bit faster.

[1] https://github.com/infinispan/infinispan/blob/master/server/hotrod/src/main/java/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.java#L49
[2] https://github.com/infinispan/infinispan/blob/master/server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java#L286

Reply:
@wburns OK, I still don't see the benefit but I can restructure the handlers as you suggest.

Reply:
If you find no benefit, don't worry about it. I was just saying that is how normally netty recommends doing this kind of stuff, with a separate decode and handler step [1]

[1] http://netty.io/wiki/user-guide-for-4.x.html#speaking-in-pojo-instead-of-bytebuf